File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char,char> umap;
        umap['}']='{';
        umap[')']='(';
        umap[']']='[';
        stack<char> st;
        for(int i=0;i<s.size();i++){
            if(umap.find(s[i])!=umap.end()){
                if(st.empty()) return false;
                if(st.top()!=umap[s[i]]) return false;
                st.pop();
            }else{
                st.push(s[i]);
            }
        }
        return st.empty();
    }
};

File: 20-valid-parentheses/valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
        Stack<Character> st=new Stack<>();

        HashMap<Character,Character> hashmap=new HashMap<>(3);
        hashmap.put(')','(');
        hashmap.put('}','{');
        hashmap.put(']','[');

        for(int i=0;i<s.length();i++){
            char ch=s.charAt(i);
            if(hashmap.containsKey(ch)){
                if(st.empty() || st.peek()!=hashmap.get(ch)) return false;
                st.pop();
            }else{
                st.push(ch);
            }
        }
        return st.empty();
    }
}